apps/sms/views/shared/test/unit/system_message_handler_test.js

```diff
@@ -0,0 +1,778 @@
+        });
+
+        test('if message has subject', function(done) {
+          mmsMessage.subject = 'message-subject';
+
+          SystemMessageHandler.onSmsReceivedSystemMessage(mmsMessage).then(
+            () => {
+              assertUserIsNotNotified();
+
+              sinon.assert.calledWith(
+                NotificationHelper.send,
+                { raw: mmsMessage.sender },
+                {
+                  icon: 'iconURI',
+                  bodyL10n: { raw: 'message-subject' },
+                  tag: 'threadId:' + mmsMessage.threadId,
+                  data: { id: mmsMessage.id, threadId: mmsMessage.threadId },
+                  closeOnClick: false
+                }
+              );
+            }
+          ).then(done, done);
+        });
+
+        test('if message does not have subject, but has text slide',
+        function(done) {
+          this.sinon.stub(SMIL, 'parse').withArgs(mmsMessage).returns(
+            Promise.resolve([{ text: 'some text' }])
+          );
+
+          SystemMessageHandler.onSmsReceivedSystemMessage(mmsMessage).then(
+            () => {
+              assertUserIsNotNotified();
+
+              sinon.assert.calledWith(
+                NotificationHelper.send,
+                { raw: mmsMessage.sender },
+                {
+                  icon: 'iconURI',
+                  bodyL10n: { raw: 'some text' },
+                  tag: 'threadId:' + mmsMessage.threadId,
+                  data: { id: mmsMessage.id, threadId: mmsMessage.threadId },
+                  closeOnClick: false
+                }
+              );
+            }
+          ).then(done, done);
+        });
+
+        test('if message does not have subject, nor text slide',
+        function(done) {
+          SystemMessageHandler.onSmsReceivedSystemMessage(mmsMessage).then(
+            () => {
+              assertUserIsNotNotified();
+
+              sinon.assert.calledWith(
+                NotificationHelper.send,
+                { raw: mmsMessage.sender },
+                {
+                  icon: 'iconURI',
+                  bodyL10n: 'mms-message',
+                  tag: 'threadId:' + mmsMessage.threadId,
+                  data: { id: mmsMessage.id, threadId: mmsMessage.threadId },
+                  closeOnClick: false
+                }
+              );
+            }
+          ).then(done, done);
+        });
+      });
+
+      suite('handling notification click', function() {
+        var notificationClickHandler;
+
+        setup(function(done) {
+          this.sinon.stub(MessageManager, 'getMessage').withArgs(
+            smsMessage.id
+          ).returns(smsMessage);
+
+          SystemMessageHandler.onSmsReceivedSystemMessage(smsMessage).then(
+            () => {
+              // We don't want to use "yield" here because we'd like to have
+              // promise-based result of yielded function.
+              notificationClickHandler = notificationStub.addEventListener.
```

Comment:
I don't know, but you could maybe spy the function and get its result ?

The idea is that normally when you "yield" you act on this stub, but getting the result of the yielded function is something internal that you should not care about.
But maybe in this case where we really need it, then you're interested in the function itself, and you should spy it.

Does it make sense ?